### PR TITLE
Add/decorators python

### PR DESCRIPTION
**Pull request by Goku-kun**
### Description

This pull request addresses the issue #211 and creates a new entry for decorators. This pull request also updates the tags to add the new 'Decorators' tag in the tags.md file.

Files added are: 
1. 'content/python/concepts/decorators/decorators.md'

Files updated are:
1. 'documentation/tags.md'

### Type of Change

- [✅] Adding a new entry
- [ ] Editing an existing entry (fixing a typo, bug, issues, etc)
- [✅] Updating the documentation

### Checklist

- [✅] My entry follows the Codecademy Docs style guide
- [✅] I have performed a self-review of my own writing and code
- [✅] I have made corresponding changes to the documentation
- [✅] My changes generate no new warnings
- [✅] I have checked my entry and corrected any misspellings
- [✅] All writings are my own

### CODE REVIEW

**Comment by Goku-kun** [comment]
Hey @sonnynomnom 

There are other markdown files under 'content/html/concepts/semantic-html/terms' which have linting errors due to extra spaces at the end of lines.

I've removed those trailing white spaces and updated the following files:

1. content/html/concepts/semantic-html/terms/details/details.md
2. content/html/concepts/semantic-html/terms/summary/summary.md

The following will be the final pull request with the above given files updated as well.

File: content/python/concepts/decorators/decorators.md [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+        original_function()
+
+        # code can be executed after the original function call
+
+    # (IMPORTANT)
+    return wrapper
+
+```
+
+The `original_function()` in this case, is the function that is decorated with the `decorator_name()`.
```
Comment (lukeecart) [inline content/python/concepts/decorators/decorators.md:46]:
```suggestion
The `original_function()`, in this case, is the function that is decorated with the `decorator_name()`.
```

File: content/python/concepts/decorators/decorators.md [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+        print(f"'{original_function.__name__}' function execution will start now.")
+
+        # Original function call with the upper cased argument
+        result = original_function(arg1)
+
+        # Code can be executed after the original function
+        result = result.upper()
+        return result
+
+    # returning the wrapper function
```
Comment (lukeecart) [inline content/python/concepts/decorators/decorators.md:68]:
```suggestion
    # Return the wrapper function
```

File: content/python/concepts/decorators/decorators.md [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+
+The general syntax for creating a decorator using classes is as follows:
+
+```py
+class decorator:
+
+    def __init__(self, original_function):
+        self.original_function = original_function
+
+
```
Comment (lukeecart) [inline content/python/concepts/decorators/decorators.md:83]: Please could we remove this extra line and in a few other places where there are doubles lines

File: content/python/concepts/decorators/decorators.md [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+def double_greeting(arg1, arg2):
+    return arg1 + ". " + arg2 + "!"
+
+
+print(greeting('hi'))
+# output:
+# HI!
+
+print(double_greeting('hi', 'hello'))
+# output:
```
Comment (lukeecart) [inline content/python/concepts/decorators/decorators.md:286]: This occurs a few times. Please could we capitalize the starts of the comments.
```suggestion
# Output:
```

File: content/python/concepts/decorators/decorators.md [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+# HI. HELLO!
+
+```
+
+## Passing Arguments to Decorators
+
+Arguments can also be passed to the decorator itself. To pass arguments to the Decorator, there should be a `decorator_wrapper_to_accept_decorator_arguments()` function wrapping the original decorator.
```
Comment (lukeecart) [inline content/python/concepts/decorators/decorators.md:293]:
```suggestion
Arguments can also be passed to the decorator itself. To pass arguments to the decorator, there should be a `decorator_wrapper_to_accept_decorator_arguments()` function wrapping the original decorator.
```

File: content/python/concepts/decorators/decorators.md [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+
+## `help()` with Decorators
+
+The original function name and docstring are lost when it's wrapped by a decorator.
+
+```codebyte/py
+def upper_case_decorator(original_function):
+
+    def wrapper():
+        """decorates a function by uppercasing the return value"""
```
Comment (lukeecart) [inline content/python/concepts/decorators/decorators.md:362]: Could we change this to a single line comment? I believe the `"""` syntax is used for multi line comments.
```suggestion
        # decorates a function by uppercasing the return value
```

File: content/python/concepts/decorators/decorators.md [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+
+        result = original_function()
+        result = result.upper()
+        return result
+
+    return wrapper
+
+@upper_case_decorator
+def greeting():
+    """ returns a str greeting """
```
Comment (lukeecart) [inline content/python/concepts/decorators/decorators.md:372]: same here. Single line comments

File: content/python/concepts/decorators/decorators.md [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+```codebyte/py
+import functools
+import time
+
+
+def performance_measure(original_function):
+
+    @functools.wraps(original_function)
+    def wrapper(loop_val):
+        """ decorates a function by measuring time for it's execution"""
```
Comment (lukeecart) [inline content/python/concepts/decorators/decorators.md:429]: Please could this be a single line comment

**Review by lukeecart — CHANGES_REQUESTED** [review]
I'm perhaps being a bit picky. But there are a few comments such as capitalizing the first word of a comment and changing`"""` comments into single-line comments. Other than that you've included some good examples and helpful information.

File: content/python/concepts/decorators/decorators.md (hunk 3 above)
Comment (Goku-kun) [reply content/python/concepts/decorators/decorators.md:83]: I believe these lines are very important to fully understand the basic syntax for a class based decorator and hence, they should stay here.

File: content/python/concepts/decorators/decorators.md (hunk 6 above)
Comment (Goku-kun) [reply content/python/concepts/decorators/decorators.md:362]: This is not a single line comment. That's not what the purpose of `"""` is. They represent docstrings in python and not multi-line comments. Using the triple double quotes, programmers can write docstrings in python to associate documentation with the code. Docstrings can also be single-line or multi-line. I'm resolving this change and cascading it to other same suggestions as well.

**Review by Goku-kun — COMMENTED** [review]
Hi @lukeecart. Thank you for reviewing the documentation and making the suggestions.
I've reviewed your suggestions and noticed the silly grammatical mistakes. I'll be updating those.

File: content/python/concepts/decorators/decorators.md (hunk 6 above)
Comment (lukeecart) [reply content/python/concepts/decorators/decorators.md:362]: @Goku-kun Fair point. I forgot about that. It was late

Comment (Goku-kun) [reply content/python/concepts/decorators/decorators.md:362]: No worries. We are a community. We help each other. With stuff like this, everyone comes out ahead. xD

File: content/python/concepts/decorators/decorators.md (hunk 2 above)
Comment (lukeecart) [reply content/python/concepts/decorators/decorators.md]: @Goku-kun There seems to be still a few comments that need to start with a capital letter. Please could we update these

**Review by sonnynomnom — APPROVED** [review]
Merging! And then going to make some minor changes and follow up in a comment.

**Comment by sonnynomnom** [comment]
A lot of changes to be made. 

Here's my first run: https://github.com/Codecademy/docs/commit/4af6085ec0fbcc1097f00c84d003a58242a2f558

**Comment by Goku-kun** [comment]
Hey @sonnynomnom.

I was following the [Curriculum Documentation Guide](http://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) and that's why I marked the keywords with italics. Am I not supposed to do that?

Also, do you want me to handle the rest of the changes that you mentioned above?

**Pull request merged**